README.md
# Luzifer / rsyslog\_cron

This project is a quick and dirty replacement for running a cron daemon inside docker containers.

## Advantages

- It logs the output of the jobs into a remote syslog target (like Papertrail) using TCP syslog
- Crons can be started on seconds, not only on minutes like a conventional cron
- Due to the logs cron jobs can get debugged
- On success and failure a HTTP ping to [Healthchecks](https://healthchecks.io/) or [Cronitor](https://cronitor.io/) can be executed

## Usage

1. Put the [binary](https://github.com/Luzifer/rsyslog_cron/releases/latest) into your container
2. Generate a YAML file containing the cron definition
3. Watch your crons get executed in your log stream

## Config format

```yaml
---
rsyslog_target: logs.myserver.com:12345
log_format: '<{{ syslogpri .Severity }}>{{ .Date.Format "Jan 02 15:04:05" }} {{ .Hostname }} {{ .JobName }}: {{ .Message }}'
jobs:
  - name: date
    schedule: "0 * * * * *"
    cmd: "/bin/date"
    args:
      - "+%+"
    ping_success: "https://..."
    ping_failure: "https://..."
```

- `rsyslog_target` - needs to be a rsyslog endpoint supporting TCP plain connections like Loggly or Papertrail does.
- `log_format` - format to use for generating the log line (above shown is default and does not need to be provided)
- `schedule` - consists of 6 instead of the normal 5 fields:

```
field         allowed values
-----         --------------
second        0-59
minute        0-59
hour          0-23
day of month  1-31
month         1-12 (or names, see below)
day of week   0-7 (0 or 7 is Sun, or use names)
```

Standard format for crontab entries is supported. (See `man 5 crontab`)


History.md
# 0.3.0 / 2017-09-13

  * Add HTTP-GET ping for success / failure
  * Deps: Move to dep instead of godeps
  * Deps: Update dependencies

# 0.2.1 / 2017-04-02

  * Fix go-vet error

# 0.2.0 / 2017-04-02

  * Add building for Github releases
  * Update parameters in README
  * Vendor dependencies
  * Add meta files
  * Allow log format to be configured


0.1.0 / 2016-02-26
==================

  - Initital version


syslog.go
package main

import (
	"bytes"
	"fmt"
	"io"
	"net"
	"text/template"
	"time"

	"github.com/cenkalti/backoff"
)

const (
	readWriteTimeout = 1 * time.Second
	tcpDialTimeout   = 5 * time.Second
)

func NewSyslogAdapter(address, logFormat string) (*SyslogAdapter, error) {
	return &SyslogAdapter{
		address:   address,
		logFormat: logFormat,
		dialer: &net.Dialer{
			Timeout: tcpDialTimeout,
		},
	}, nil
}

type message struct {
	Date     time.Time
	JobName  string
	Message  string
	Severity int
}

type SyslogAdapter struct {
	address   string
	logFormat string
	dialer    *net.Dialer
}

func (a *SyslogAdapter) formatMessage(m *message) (string, error) {
	t, err := template.New("logFormat").Funcs(template.FuncMap{
		"syslogpri": func(pri int) int { return 16*8 + pri },
	}).Parse(a.logFormat)

	if err != nil {
		return "", err
	}

	buf := bytes.NewBufferString("")
	if err := t.Execute(buf, map[string]interface{}{
		"Date":     m.Date,
		"JobName":  m.JobName,
		"Hostname": cfg.Hostname,
		"Message":  m.Message,
		"Severity": m.Severity,
	}); err != nil {
		return "", err
	}

	return buf.String(), nil
}

func (a *SyslogAdapter) Stream(logstream chan *message) {
	backoff.Retry(func() error {
		conn, err := a.dialer.Dial("tcp", a.address)
		if err != nil {
			fmt.Printf("syslog: Unable to dial to remote address\n")
			return fmt.Errorf("Catch me if you can.")
		}
		defer conn.Close()

		b := new(bytes.Buffer)
		for msg := range logstream {
			b.Reset()

			msgLine, err := a.formatMessage(msg)
			if err != nil {
				return err
			}
			fmt.Fprintln(b, msgLine)

			if err := conn.SetDeadline(time.Now().Add(readWriteTimeout)); err != nil {
				fmt.Printf("syslog: Unable to set deadline: %s\n", err)
				return fmt.Errorf("Catch me if you can.")
			}

			logLine := b.Bytes()
			written, err := io.Copy(conn, b)

			if err != nil {
				if written > 0 {
					fmt.Printf("syslog: (%d/%d) %s\n", written, len(logLine), err)
				} else {
					fmt.Printf("syslog: %s\n", err)
				}
				return fmt.Errorf("syslog: %s", err)
			}
		}

		fmt.Printf("syslog: I got out of the channel watch. This should never happen.\n")
		return fmt.Errorf("Wat? Why am I here?")

	}, &backoff.ZeroBackOff{})
}


main.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"os/exec"
	"strings"
	"time"

	"github.com/Luzifer/rconfig"
	"github.com/robfig/cron"
	"golang.org/x/net/context"
	"gopkg.in/yaml.v2"
)

var (
	cfg = struct {
		ConfigFile string        `flag:"config" default:"config.yaml" description:"Cron definition file"`
		Hostname   string        `flag:"hostname" description:"Overwrite system hostname"`
		PingTimout time.Duration `flag:"ping-timeout" default:"1s" description:"Timeout for success / failure pings"`
	}{}
	version = "dev"

	logstream = make(chan *message, 1000)
)

type cronConfig struct {
	RSyslogTarget string    `yaml:"rsyslog_target"`
	LogTemplate   string    `yaml:"log_template"`
	Jobs          []cronJob `yaml:"jobs"`
}

type cronJob struct {
	Name        string   `yaml:"name"`
	Schedule    string   `yaml:"schedule"`
	Command     string   `yaml:"cmd"`
	Arguments   []string `yaml:"args"`
	PingSuccess string   `yaml:"ping_success"`
	PingFailure string   `yaml:"ping_failure"`
}

func init() {
	rconfig.Parse(&cfg)

	if cfg.Hostname == "" {
		hostname, _ := os.Hostname()
		cfg.Hostname = hostname
	}
}

func main() {
	body, err := ioutil.ReadFile(cfg.ConfigFile)
	if err != nil {
		log.Fatalf("Unable to read config file: %s", err)
	}

	cc := cronConfig{
		LogTemplate: `<{{ syslogpri .Severity }}>{{ .Date.Format "Jan 02 15:04:05" }} {{ .Hostname }} {{ .JobName }}: {{ .Message }}`,
	}
	if err := yaml.Unmarshal(body, &cc); err != nil {
		log.Fatalf("Unable to parse config file: %s", err)
	}

	c := cron.New()

	for i := range cc.Jobs {
		job := cc.Jobs[i]
		if err := c.AddFunc(job.Schedule, getJobExecutor(job)); err != nil {
			log.Fatalf("Unable to add job '%s': %s", job.Name, err)
		}
	}

	c.Start()

	logadapter, err := NewSyslogAdapter(cc.RSyslogTarget, cc.LogTemplate)
	if err != nil {
		log.Fatalf("Unable to open syslog connection: %s", err)
	}
	logadapter.Stream(logstream)
}

func getJobExecutor(job cronJob) func() {
	return func() {
		stdout := &messageChanWriter{
			jobName:  job.Name,
			msgChan:  logstream,
			severity: 6, // Informational
		}

		stderr := &messageChanWriter{
			jobName:  job.Name,
			msgChan:  logstream,
			severity: 3, // Error
		}

		fmt.Fprintln(stdout, "[SYS] Starting job")

		cmd := exec.Command(job.Command, job.Arguments...)
		cmd.Stdout = stdout
		cmd.Stderr = stderr

		err := cmd.Run()
		switch err.(type) {
		case nil:
			fmt.Fprintln(stdout, "[SYS] Command execution successful")
			go func(url string) {
				if err := doPing(url); err != nil {
					fmt.Fprintf(stderr, "[SYS] Ping to URL %q caused an error: %s", url, err)
				}
			}(job.PingSuccess)

		case *exec.ExitError:
			fmt.Fprintln(stderr, "[SYS] Command exited with unexpected exit code != 0")
			go func(url string) {
				if err := doPing(url); err != nil {
					fmt.Fprintf(stderr, "[SYS] Ping to URL %q caused an error: %s", url, err)
				}
			}(job.PingFailure)

		default:
			fmt.Fprintf(stderr, "[SYS] Execution caused error: %s\n", err)
			go func(url string) {
				if err := doPing(url); err != nil {
					fmt.Fprintf(stderr, "[SYS] Ping to URL %q caused an error: %s", url, err)
				}
			}(job.PingFailure)

		}
	}
}

func doPing(url string) error {
	if url == "" {
		return nil
	}

	ctx, cancel := context.WithTimeout(context.Background(), cfg.PingTimout)
	defer cancel()

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return err
	}
	resp, err := http.DefaultClient.Do(req.WithContext(ctx))
	if err != nil {
		return err
	}

	if resp.StatusCode > 299 {
		return fmt.Errorf("Expected HTTP2xx status, got HTTP%d", resp.StatusCode)
	}

	return nil
}

type messageChanWriter struct {
	jobName  string
	msgChan  chan *message
	severity int

	buffer []byte
}

func (m *messageChanWriter) Write(p []byte) (n int, err error) {
	n = len(p)
	err = nil

	m.buffer = append(m.buffer, p...)
	if strings.Contains(string(m.buffer), "\n") {
		lines := strings.Split(string(m.buffer), "\n")
		for _, l := range lines[:len(lines)-1] {
			m.msgChan <- &message{
				Date:     time.Now(),
				JobName:  m.jobName,
				Message:  l,
				Severity: m.severity,
			}
		}
		m.buffer = []byte(lines[len(lines)-1])
	}

	return
}
